apps/ejabberd/src/mod_muc_db_mnesia.erl

```diff
@@ -0,0 +1,239 @@
+%%%----------------------------------------------------------------------
+%%% File    : mod_muc.erl
+%%% Author  : Alexey Shchepin <alexey@process-one.net>
+%%% Purpose : MUC support (XEP-0045)
+%%% Created : 19 Mar 2003 by Alexey Shchepin <alexey@process-one.net>
+%%%
+%%%
+%%% ejabberd, Copyright (C) 2002-2011   ProcessOne
+%%%
+%%% This program is free software; you can redistribute it and/or
+%%% modify it under the terms of the GNU General Public License as
+%%% published by the Free Software Foundation; either version 2 of the
+%%% License, or (at your option) any later version.
+%%%
+%%% This program is distributed in the hope that it will be useful,
+%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
+%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+%%% General Public License for more details.
+%%%
+%%% You should have received a copy of the GNU General Public License
+%%% along with this program; if not, write to the Free Software
+%%% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+%%% 02111-1307 USA
+%%%
+%%%----------------------------------------------------------------------
+
+-module(mod_muc_db_mnesia).
+-behaviour(mod_muc_db).
+-export([init/2,
+         store_room/4,
+         restore_room/3,
+         forget_room/3,
+         get_rooms/2,
+         can_use_nick/4,
+         get_nick/3,
+         set_nick/4]).
+
+-include("ejabberd.hrl").
+-include("jlib.hrl").
+-include_lib("ejabberd/include/mod_muc.hrl").
+
+init(Host, Opts) ->
+    mnesia:create_table(muc_room,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_room)}]),
+    mnesia:create_table(muc_registered,
+                        [{disc_copies, [node()]},
+                         {attributes, record_info(fields, muc_registered)}]),
+    mnesia:add_table_copy(muc_room, node(), disc_copies),
+    mnesia:add_table_copy(muc_registered, node(), disc_copies),
+    MyHost = gen_mod:get_opt_subhost(Host, Opts, <<"conference.@HOST@">>),
+    update_tables(MyHost),
+    mnesia:add_table_index(muc_registered, nick),
+    ok.
+
+-spec store_room(ejabberd:server(), ejabberd:server(), mod_muc:room(), list())
+            -> {'aborted',_} | {'atomic',_}.
+store_room(_LServer, Host, Name, Opts) ->
+    F = fun() ->
+                mnesia:write(#muc_room{name_host = {Name, Host},
+                                       opts = Opts})
+        end,
+    mnesia:transaction(F).
+
+
+-spec restore_room(ejabberd:server(), ejabberd:server(), mod_muc:room())
+                                    -> 'error' | 'undefined' | [any()].
+restore_room(_LServer, Host, Name) ->
+    case catch mnesia:dirty_read(muc_room, {Name, Host}) of
+        [#muc_room{opts = Opts}] ->
+            Opts;
+        _ ->
+            error
```

Comment:
Although `drity_read` will abort with an exception on some Mnesia error, I think we should have more meaningful error tuple here for empty result.